src/test/java/teammates/test/cases/newaction/BaseActionTest.java

```diff
@@ -0,0 +1,271 @@
+package teammates.test.cases.newaction;
+
+import java.util.List;
+import java.util.Map;
+
+import org.testng.annotations.BeforeClass;
+
+import teammates.common.datatransfer.DataBundle;
+import teammates.common.datatransfer.UserInfo;
+import teammates.common.datatransfer.attributes.InstructorAttributes;
+import teammates.common.datatransfer.attributes.StudentAttributes;
+import teammates.common.exception.InvalidHttpParameterException;
+import teammates.common.exception.UnauthorizedAccessException;
+import teammates.common.util.EmailWrapper;
+import teammates.test.cases.BaseComponentTestCase;
+import teammates.ui.newcontroller.Action;
+import teammates.ui.newcontroller.JsonResult;
+
+/**
+ * Base class for all action tests.
+ *
+ * @param <T> The action class being tested.
+ */
+public abstract class BaseActionTest<T extends Action> extends BaseComponentTestCase {
+
+    protected DataBundle typicalBundle = getTypicalDataBundle();
+
+    protected abstract String getActionUri();
+
+    protected abstract String getRequestMethod();
+
+    @SuppressWarnings("unchecked")
+    protected T getAction(String... params) {
+        return (T) gaeSimulation.getNewActionObject(getActionUri(), getRequestMethod(), params);
+    }
+
+    @BeforeClass
+    public void baseClassSetup() {
```

Comment:
Make it as `@BeforeMethod`?

Reply:
Any good reason for this?

Reply:
So there is no state leak between `@Test`? Writing all tests inside `testExecute` would result in couplings between test case. Hard to add/change test cases.

In action test level, doing `removeAndRestoreDataBundle` is really cheap and fast unlink end-to-end test.

Reply:
Makes sense. Made the change.